File: runners/dotnet/Support/TestContext.cs
using OpenFga.Sdk.Client;
using OpenFga.Sdk.Client.Model;
using OpenFga.Sdk.Model;
using FluentAssertions;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace OpenFga.Sdk.Conformance.Support
{
    public class TestContext
    {
        public OpenFgaClient? Client { get; set; }
        public ClientConfiguration? Config { get; set; }
        public object? LastResponse { get; set; }
        public Exception? LastError { get; set; }
        public Dictionary<string, object> SavedData { get; set; } = new();
        public Dictionary<string, string> Headers { get; set; } = new();
        public Dictionary<string, string>? ResponseHeaders { get; set; }
        public Dictionary<string, string>? AuthConfig { get; set; }
        public int? StatusCode { get; set; }
        public string WiremockUrl { get; set; } = "http://localhost:8080";

        public void Reset()
        {
            Client = null;
            Config = null;
            LastResponse = null;
            LastError = null;
            SavedData.Clear();
            Headers.Clear();
            ResponseHeaders = null;
            AuthConfig = null;
            StatusCode = null;
        }

        public async Task CreateClientAsync(ClientConfiguration config)
        {
            try
            {
                Config = new ClientConfiguration
                {
                    ApiUrl = WiremockUrl,
                    StoreId = config.StoreId,
                    AuthorizationModelId = config.AuthorizationModelId,
                    Credentials = config.Credentials
                };

                Client = new OpenFgaClient(Config);
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Failed to create OpenFGA client: {ex.Message}", ex);
            }
        }

        public async Task ExecuteApiCallAsync(Func<Task<object>> apiCall)
        {
            try
            {
                LastResponse = await apiCall();
                LastError = null;
            }
            catch (Exception ex)
            {
                LastError = ex;
                LastResponse = null;
                
                // Enhanced error handling with status code extraction
                ExtractStatusCodeFromException(ex);
            }
        }

        private void ExtractStatusCodeFromException(Exception ex)
        {
            // Extract HTTP status code from various exception types
            if (ex is HttpRequestException httpEx)
            {
                // Try to extract status code from message or data
                if (ex.Data.Contains("StatusCode"))
                {
                    StatusCode = (int)ex.Data["StatusCode"];
                }
                else if (httpEx.Message.Contains("401"))
                {
                    StatusCode = 401;
                }
                else if (httpEx.Message.Contains("403"))
                {
                    StatusCode = 403;
                }
                else if (httpEx.Message.Contains("404"))
                {
                    StatusCode = 404;
                }
                else if (httpEx.Message.Contains("400"))
                {
                    StatusCode = 400;
                }
            }
            // Add other SDK-specific exception handling as needed
        }

        // Enhanced authentication configuration
        public void ConfigureAuthentication(string method, Dictionary<string, string> config)
        {
            switch (method.ToLowerInvariant())
            {
                case "bearer":
                    Headers["Authorization"] = $"Bearer {config["token"]}";
                    break;
                case "client_credentials":
                    // Store credentials for token exchange
                    AuthConfig = new Dictionary<string, string>
                    {
                        ["method"] = "client_credentials",
                        ["clientId"] = config.GetValueOrDefault("clientId", ""),
                        ["clientSecret"] = config.GetValueOrDefault("clientSecret", ""),
                        ["apiTokenIssuer"] = config.GetValueOrDefault("apiTokenIssuer", ""),
                        ["apiAudience"] = config.GetValueOrDefault("apiAudience", "")
                    };
                    break;
                default:
                    throw new ArgumentException($"Unsupported authentication method: {method}");
            }
        }

        // Response header inspection
        public string GetResponseHeader(string headerName)
        {
            if (ResponseHeaders != null && ResponseHeaders.ContainsKey(headerName))
            {
                return ResponseHeaders[headerName];
            }
            return null;
        }

        // Enhanced status code checking
        public int? GetStatusCode()
        {
            if (StatusCode.HasValue)
            {
                return StatusCode.Value;
            }
            // Assume 200 for successful responses
            return LastResponse != null ? 200 : null;
        }

        // Assertion helpers
        public void AssertResponseSuccess()
        {
            if (LastError != null)
            {
                throw new AssertionException($"Expected successful response, got error: {LastError.Message}");
            }
            LastResponse.Should().NotBeNull();
        }

        public void AssertResponseFailure()
        {
            if (LastError == null)
            {
                throw new AssertionException("Expected error response, but got success");
            }
        }

        public void AssertResponseContainsItems(int count)
        {
            AssertResponseSuccess();
            
            if (LastResponse is ReadResponse readResponse)
            {
                var actualCount = readResponse.Tuples?.Count ?? 0;
                actualCount.Should().Be(count, $"Expected {count} items, got {actualCount}");
            }
            else
            {
                throw new AssertionException("Response is not a ReadResponse");
            }
        }

        public void AssertTuplesInclude(List<TupleKey> expectedTuples)
        {
            AssertResponseSuccess();
            
            if (LastResponse is ReadResponse readResponse)
            {
                var actualTuples = readResponse.Tuples ?? new List<Tuple>();
                
                foreach (var expected in expectedTuples)
                {
                    var found = actualTuples.Any(tuple =>
                        tuple.Key.User == expected.User &&
                        tuple.Key.Relation == expected.Relation &&
                        tuple.Key.Object == expected.Object);
                    
                    if (!found)
                    {
                        throw new AssertionException($"Expected tuple not found: user={expected.User}, relation={expected.Relation}, object={expected.Object}");
                    }
                }
            }
            else
            {
                throw new AssertionException("Response is not a ReadResponse");
            }
        }

        // Save and compare responses
        public void SaveResponse(string key)
        {
            if (LastResponse == null)
            {
                throw new InvalidOperationException("No response to save");
            }
            SavedData[key] = LastResponse;
        }

        public void CompareWithSaved(string key, bool shouldMatch = true)
        {
            if (!SavedData.TryGetValue(key, out var savedData))
            {
                throw new InvalidOperationException($"No saved data found for key: {key}");
            }
            
            if (LastResponse == null)
            {
                throw new InvalidOperationException("No current response to compare");
            }

            var currentJson = System.Text.Json.JsonSerializer.Serialize(LastResponse);
            var savedJson = System.Text.Json.JsonSerializer.Serialize(savedData);
            var matches = currentJson == savedJson;

            if (shouldMatch && !matches)
            {
                throw new AssertionException($"Current response does not match saved response for key: {key}");
            }
            else if (!shouldMatch && matches)
            {
                throw new AssertionException($"Current response matches saved response for key: {key} (expected different)");
            }
        }
    }

    public class AssertionException : Exception
    {
        public AssertionException(string message) : base(message) { }
    }
}

File: runners/java/src/test/java/dev/openfga/sdk/conformance/support/TestContext.java
package dev.openfga.sdk.conformance.support;

import dev.openfga.sdk.api.client.OpenFgaClient;
import dev.openfga.sdk.api.configuration.ClientConfiguration;
import dev.openfga.sdk.api.model.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.assertj.core.api.Assertions;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class TestContext {
    private OpenFgaClient client;
    private ClientConfiguration config;
    private Object lastResponse;
    private Exception lastError;
    private Map<String, Object> savedData = new HashMap<>();
    private Map<String, String> headers = new HashMap<>();
    private Map<String, String> responseHeaders = new HashMap<>();
    private Map<String, String> authConfig = new HashMap<>();
    private Integer statusCode;
    private String wiremockUrl = System.getenv().getOrDefault("WIREMOCK_URL", "http://localhost:8080");
    private ObjectMapper objectMapper = new ObjectMapper();

    public void reset() {
        client = null;
        config = null;
        lastResponse = null;
        lastError = null;
        savedData.clear();
        headers.clear();
        responseHeaders.clear();
        authConfig.clear();
        statusCode = null;
    }

    public void createClient(Map<String, String> configMap) throws Exception {
        try {
            ClientConfiguration.Builder configBuilder = new ClientConfiguration.Builder()
                    .apiUrl(wiremockUrl);
            
            if (configMap.containsKey("storeId")) {
                configBuilder.storeId(configMap.get("storeId"));
            }
            
            if (configMap.containsKey("authorizationModelId")) {
                configBuilder.authorizationModelId(configMap.get("authorizationModelId"));
            }
            
            config = configBuilder.build();
            client = new OpenFgaClient(config);
        } catch (Exception e) {
            throw new Exception("Failed to create OpenFGA client: " + e.getMessage(), e);
        }
    }

    public void executeApiCall(ApiCall apiCall) {
        try {
            lastResponse = apiCall.call();
            lastError = null;
        } catch (Exception e) {
            lastError = e;
            lastResponse = null;
            
            // Enhanced error handling with status code extraction
            extractStatusCodeFromException(e);
        }
    }

    private void extractStatusCodeFromException(Exception ex) {
        // Extract HTTP status code from various exception types
        String errorMessage = ex.getMessage().toLowerCase();
        
        if (errorMessage.contains("401") || errorMessage.contains("unauthorized")) {
            statusCode = 401;
        } else if (errorMessage.contains("403") || errorMessage.contains("forbidden")) {
            statusCode = 403;
        } else if (errorMessage.contains("404") || errorMessage.contains("not found")) {
            statusCode = 404;
        } else if (errorMessage.contains("400") || errorMessage.contains("bad request")) {
            statusCode = 400;
        }
        
        // Try to extract from exception data or cause
        if (ex.getCause() != null) {
            extractStatusCodeFromException(ex.getCause());
        }
    }

    // Enhanced authentication configuration
    public void configureAuthentication(String method, Map<String, String> config) {
        switch (method.toLowerCase()) {
            case "bearer":
                headers.put("Authorization", "Bearer " + config.get("token"));
                break;
            case "client_credentials":
                // Store credentials for token exchange
                authConfig.put("method", "client_credentials");
                authConfig.put("clientId", config.getOrDefault("clientId", ""));
                authConfig.put("clientSecret", config.getOrDefault("clientSecret", ""));
                authConfig.put("apiTokenIssuer", config.getOrDefault("apiTokenIssuer", ""));
                authConfig.put("apiAudience", config.getOrDefault("apiAudience", ""));
                break;
            default:
                throw new IllegalArgumentException("Unsupported authentication method: " + method);
        }
    }

    // Response header inspection
    public String getResponseHeader(String headerName) {
        return responseHeaders.get(headerName.toLowerCase());
    }

    // Enhanced status code checking
    public Integer getStatusCode() {
        if (statusCode != null) {
            return statusCode;
        }
        // Assume 200 for successful responses
        return lastResponse != null ? 200 : null;
    }

    public CompletableFuture<Void> executeAsyncApiCall(AsyncApiCall apiCall) {
        return apiCall.call()
                .thenAccept(response -> {
                    lastResponse = response;
                    lastError = null;
                })
                .exceptionally(throwable -> {
                    lastError = (Exception) throwable;
                    lastResponse = null;
                    return null;
                });
    }

    // Assertion helpers
    public void assertResponseSuccess() {
        if (lastError != null) {
            throw new AssertionError("Expected successful response, got error: " + lastError.getMessage());
        }
        Assertions.assertThat(lastResponse).isNotNull();
    }

    public void assertResponseFailure() {
        if (lastError == null) {
            throw new AssertionError("Expected error response, but got success");
        }
    }

    public void assertResponseContainsItems(int count) {
        assertResponseSuccess();
        
        if (lastResponse instanceof ReadResponse) {
            ReadResponse readResponse = (ReadResponse) lastResponse;
            int actualCount = readResponse.getTuples() != null ? readResponse.getTuples().size() : 0;
            Assertions.assertThat(actualCount).isEqualTo(count);
        } else {
            throw new AssertionError("Response is not a ReadResponse");
        }
    }

    public void assertTuplesInclude(List<Map<String, String>> expectedTuples) {
        assertResponseSuccess();
        
        if (lastResponse instanceof ReadResponse) {
            ReadResponse readResponse = (ReadResponse) lastResponse;
            List<Tuple> actualTuples = readResponse.getTuples();
            
            if (actualTuples == null) {
                throw new AssertionError("Response contains no tuples");
            }
            
            for (Map<String, String> expected : expectedTuples) {
                boolean found = actualTuples.stream().anyMatch(tuple -> 
                    tuple.getKey().getUser().equals(expected.get("user")) &&
                    tuple.getKey().getRelation().equals(expected.get("relation")) &&
                    tuple.getKey().getObject().equals(expected.get("object"))
                );
                
                if (!found) {
                    throw new AssertionError(String.format(
                        "Expected tuple not found: user=%s, relation=%s, object=%s",
                        expected.get("user"), expected.get("relation"), expected.get("object")
                    ));
                }
            }
        } else {
            throw new AssertionError("Response is not a ReadResponse");
        }
    }

    // Save and compare responses
    public void saveResponse(String key) {
        if (lastResponse == null) {
            throw new RuntimeException("No response to save");
        }
        savedData.put(key, lastResponse);
    }

    public void compareWithSaved(String key, boolean shouldMatch) {
        Object savedData = this.savedData.get(key);
        if (savedData == null) {
            throw new RuntimeException("No saved data found for key: " + key);
        }
        
        if (lastResponse == null) {
            throw new RuntimeException("No current response to compare");
        }

        try {
            String currentJson = objectMapper.writeValueAsString(lastResponse);
            String savedJson = objectMapper.writeValueAsString(savedData);
            boolean matches = currentJson.equals(savedJson);

            if (shouldMatch && !matches) {
                throw new AssertionError("Current response does not match saved response for key: " + key);
            } else if (!shouldMatch && matches) {
                throw new AssertionError("Current response matches saved response for key: " + key + " (expected different)");
            }
        } catch (Exception e) {
            throw new RuntimeException("Failed to compare responses: " + e.getMessage(), e);
        }
    }

    // Getters and setters
    public OpenFgaClient getClient() { return client; }
    public ClientConfiguration getConfig() { return config; }
    public Object getLastResponse() { return lastResponse; }
    public Exception getLastError() { return lastError; }
    public Map<String, Object> getSavedData() { return savedData; }
    public Map<String, String> getHeaders() { return headers; }
    public String getWiremockUrl() { return wiremockUrl; }

    @FunctionalInterface
    public interface ApiCall {
        Object call() throws Exception;
    }

    @FunctionalInterface
    public interface AsyncApiCall {
        CompletableFuture<Object> call();
    }
}
